src/code/z_actor.c

```diff
@@ -2707,19 +2708,88 @@ void Actor_DrawLensActors(PlayState* play, s32 numInvisibleActors, Actor** invis
-    f32 var;
+/**
+ * Tests if an actor is currently within the bounds of its own culling volume.
+ *
+ * The culling volume is a 3D shape composed of a frustum with a box attached to the end of it. The frustum sits at the
+ * camera's position and projects forward, encompassing the player's current view; the box extrudes behind the camera,
+ * allowing actors in the immediate vicinity behind and to the sides of the camera to be detected.
+ *
+ * This function returns true if the actor is within bounds, false if not.
+ * The comparison is done in projective space against the actor's projected position as the viewing frustum
+ * in world space transforms to a box in projective space, making the calculation easy.
+ *
+ * Every actor can set properties for their own culling volume, changing its dimensions to suit the needs of it and
+ * its environment. Note that the following properties are projective properties as they are compared
+ * with the actor's position after perspective projection is applied. The units are therefore not directly comparable to
+ * world units. These depend on the current view parameters (fov, aspect, scale, znear, zfar). The default parameters
+ * considered are (60 degrees, 4/3, 1.0, 10, 12800).
+ *
+ *    cullingVolumeDistance: Configures how far forward the far plane of the frustum should extend.
+ *                           This along with cullingVolumeScale determines the maximum distance from
+ *                           the camera eye that the actor can be detected at. This quantity is related
+ *                           to world units by a factor of
+ *                                   (znear - zfar) / ((znear + zfar) * scale).
+ *                           For default view parameters, increasing this property by 1 increases the
+ *                           distance by ~0.995 world units.
+ *
+ *    cullingVolumeScale: Scales the entire culling volume in all directions except the downward
+ *                        direction. Both the frustum and the box will scale in size. This quantity is
+ *                        related to world units by different factors based on direction:
+ *                         - For the forward and backward directions, they are related in the same way
+ *                           as above. For default view parameters, increasing this property by 1 increases
+ *                           the forward and backward scales by ~0.995 world units.
+ *                         - For the sideways directions, the relation to world units is
+ *                                   (aspect / scale) * tan(0.5 * fov)
+ *                           For default view parameters, increasing this property by 1 increases the
+ *                           sideways scales by ~0.77 world units.
+ *                         - For the upward direction, the relation to world units is
+ *                                   (1 / scale) * tan(0.5 * fov)
```

Comment:
for all these "relation to world units" expressions, how does the `projW < 1.0f` check below affect things? i.e. when is the scale normalized out, and when does it affect the "relation to world units" expressions here? This only matters in Hyrule Field on N64 and I've forgotten the details, but documenting it more here could have saved me some trouble

Reply:
similarly @Thar0

Reply:
The clamp on w only changes how things behave behind the camera. Specifically it's responsible for the box shape, otherwise the culling volume it would be a pyramid shape.

The scale factor never factors out.

Let $\left ( x, y, z, 1 \right )$ be the word coordinates, then the projected coordinates are:

$$X_p = \frac{s}{A} \cot \left( \frac{1}{2} F \right ) x$$
$$Y_p = s \cot \left ( \frac{1}{2} F \right ) y$$
$$Z_p = \frac{s}{Z_{near} - Z_{far}} \left( \left( Z_{near} + Z_{far} \right) z + 2 Z_{near} Z_{far} \right)$$
$$W_p = -sz$$

The inequalities that must be satisfied are ( let $w = \max(1, W_p)$ )

$$Z_p + P_{scale} > 0$$
$$Z_p < P_{distance} + P_{scale}$$
$$\left | X_p \right | < w + P_{scale}$$
$$Y_p + w + P_{downward}>0$$
$$Y_p < w + P_{scale}$$

There's no case where these are independent of $s$, the culling parameters are always scaled.

Reply:
> Specifically it's responsible for the box shape, otherwise the culling volume it would be a pyramid shape.

ah this makes it much clearer, thanks. I'd suggest putting that in a comment, something like

```
s32 Actor_CullingVolumeTest(PlayState* play, Actor* actor, Vec3f* projPos, f32 projW) {
    if ((projPos->z > -actor->cullingVolumeScale) &&
        (projPos->z < (actor->cullingVolumeDistance + actor->cullingVolumeScale))) {
        // Clamping `projW` affects points behind the camera, so that the culling volume has a pyramid
        // shape in front of the camera and a box shape behind the camera.
        f32 invW = (projW < 1.0f) ? 1.0f : 1.0f / projW;

        if ((((fabsf(projPos->x) - actor->cullingVolumeScale) * invW) < 1.0f) &&
            (((projPos->y + actor->cullingVolumeDownward) * invW) > -1.0f) &&
            (((projPos->y - actor->cullingVolumeScale) * invW) < 1.0f)) {
            return true;
        }
    }
}
```

(No idea if changing the scoping like that matches, it would be nice if it did though).